Added the glimpse network

LewisOs opened the pull request:
Hi Shane,

I've added the code for the glimpse network (this time without confused comments to myself about inputs...). I had named the layers in such a way to make it very clear what they were doing as I worked on them, but they are probably a bit ungainly. Would you like me to rename them more concisely?

Cheers, Lewis.

shanest commented on models.py:
```diff
@@ -179,21 +179,29 @@ def glimpse_network(images, locs, scope='glimpse_network'):
                     activation = tf.nn.relu)
             
+            dense_sensor_layer = tf.layers.dense(
+                    inputs = hidden_sensor_layer,
```
Minor style comment: when passing named arguments into methods like this in Python, I think it's better to not have spaces around the equals sign, i.e. `inputs=hidden_sensor_layer`.  The reason is that, with spaces, you're making an assignment to a variable, which is not quite what's happening here (passing a value with a given name).

shanest commented on models.py:
```diff
@@ -179,21 +179,29 @@ def glimpse_network(images, locs, scope='glimpse_network'):
-                    inputs = tf.concat(
-                            values = [hidden_sensor_layer, hidden_location_layer],
-                            axis = 0),
+                    inputs = tf.add(
```
`glimpse_out_layer` is also what should also be returned by the method.  I.e. the four lines of comments from `# NOTE:` and the `return` line following it should be replaced by `return glimpse_out_layer`

shanest commented:
Looks very nice over-all!  Two changes requested: one stylistic, and one about what `glimpse_network` should be returning

shanest commented:
Oh, and about variable names: I think your names are perfect!  In general, longer, more meaningful names should actually be preferred.  It helps a lot with reading code and debugging in the future.  (In fact, one of my pet peeves in a lot of machine learning is that people use variable names like `x_t`, `g_l`, etc, which are pretty hard to understand.

shanest commented on models.py:
```diff
@@ -179,21 +179,29 @@ def glimpse_network(images, locs, scope='glimpse_network'):
             # TODO: Lewis implements Glimpse Network here, using also
```
And this `TODO` comment can be deleted now 😄

LewisOs commented:
Hi Shane,

I've sorted those bits. 

I've also renamed rnn_output to rnn_state, as discussed. Am I right to think that you were actually using the state, but had called it output, so that the meat of the code doesn't need altering?

Plus, I've deleted the location network TODO, since it's already implemented.

Hope that all looks good. I'm moving onto implementing the rnn cell now.

Ta, Lewis.

shanest commented:
Hi Lewis,

All looks great!  

There is in fact one line where I wind up passing the rnn's output, instead of its state, to the location_network.  But I'll just fix that one line very quickly :-).

Looking forward to seeing the new `rnn_cell`: it should be a method that takes two inputs (the previous state of the rnn, and the new output from the glimpse) and returns an output of the right size (for reasons we can discuss soon, it should return a pair, consisting of two copies of the output: one for the output and one for the state, which are distinguished by LSTMs but will be the same for this RNN cell).

Cheers,
Shane
